fix issue 3396 - call to unimplemented abstract method lead to linkin…

SixthDot opened the pull request:
…g error

Commit:
fix issue 3396 - call to unimplemented abstract method lead to linking error

maxhaton commented on compiler/test/fail_compilation/issue3396.d:
```diff
@@ -0,0 +1,24 @@
+/*
+TEST_OUTPUT:
+---
+fail_compilation/issue3396.d(16): Error: call to unimplemented function `M`
```
Give more context in the error message (e.g. "superclass function M")

dkorpel commented on compiler/src/dmd/expressionsem.d:
```diff
@@ -5488,6 +5488,14 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
+        if (supDotFun && supDotFun.e1.isSuperExp() && exp.f && exp.f.isAbstract() && !exp.f.fbody)
+        {
+            exp.error("call to unimplemented abstract function `%s`", exp.f.toFullSignature());
+            exp.errorSupplemental("declared here: %s", exp.f.loc.toChars());
```
It's better to put the supplemental error at `f.loc` so you get there when you ctrl+click the line in VSCode (and other editors too probably)

SixthDot replied:
that wont be nicely aligned anymore then

SixthDot replied:
I can patch the loc actually

dkorpel replied:
If https://github.com/dlang/dmd/pull/15646 gets through, both need to be converted to free functions

SixthDot replied:
yes I've seen that PR earlier. But for this problem I'm no convinced. Patching the loc allows to drop the "Error:" part of the message but IRL the two messages will still be unaligned (with two different modules)

dkorpel replied:
I don't understand what you mean. Alignment of `errorSupplemental` with the preceding `error` is a general problem, not specific to this particular error. Also, setting `exp.loc = x` and calling `exp.errorSupplemental(...)` is equivalent to `.errorSupplemental(x, ...)`, so I don't see how it relates to alignment.

SixthDot replied:
Allright. I'll use `.errorSupplemental` too.

dkorpel replied:
https://github.com/dlang/dmd/pull/15646 got merged, so now you need to use `.error` as well.

Commit:
alternative version with better err msg alignment

# Conflicts:
#	compiler/src/dmd/expressionsem.d

SixthDot commented:
thanks for the notice @dkorpel, rebased accordingly. 

Not that since the day you reviewed I've added a second commit.

- the first matches to your suggestion
- the second matches to the very first commit with a second fname in the supplemental (and better alignment in the console)

So that either only the first or both can be merged, depending on personal tastes. (access to my fork is not blocked, so you can drop the second or fixup the first)

thewilsonator commented on compiler/test/fail_compilation/issue3396.d:
```diff
@@ -0,0 +1,25 @@
+    override void M(){ super.M(); }
+}
+
+int main()
+{
+    auto b=new B();
+    b.M();
+    return 0;
+}
```
please remove this, if a function that uses them is required to reproduce the bug it should not be called `main`. Otherwise this looks good.

RazvanN7 replied:
What's the problem with it being named "main"?

SixthDot replied:
done
